lib/ddtrace/span.rb

```diff
@@ -247,13 +249,71 @@ def to_hash
       }
 
       if !@start_time.nil? && !@end_time.nil?
-        h[:start] = (@start_time.to_f * 1e9).to_i
-        h[:duration] = ((@end_time - @start_time) * 1e9).to_i
+        h[:start] = start_time_nano
+        h[:duration] = duration_nano
       end
 
       h
     end
 
+    # MessagePack serializer interface. Making this object
+    # respond to `#to_msgpack` allows it to be automatically
+    # serialized by MessagePack.
+    #
+    # This is more efficient than doing +MessagePack.pack(span.to_hash)+
+    # as we don't have to create an intermediate Hash.
+    #
+    # @param packer [MessagePack::Packer] serialization buffer, can be +nil+ with JRuby
+    def to_msgpack(packer = nil)
+      # As of 1.3.3, JRuby implementation doesn't pass an existing packer
+      packer ||= MessagePack::Packer.new
+
+      if !@start_time.nil? && !@end_time.nil?
+        packer.write_map_header(13) # Set header with how many elements in the map
+
+        packer.write('start')
+        packer.write(start_time_nano)
+
+        packer.write('duration')
+        packer.write(duration_nano)
+      else
+        packer.write_map_header(11) # Set header with how many elements in the map
+      end
+
+      # DEV: We use strings as keys here, instead of symbols, as
+      # DEV: MessagePack will ultimately convert them to strings.
+      # DEV: By providing strings directly, we skip this indirection operation.
```

Comment:
This would be faster than defining a bunch of constants, ie 

```
SPAN_ID = 'span_id'.freeze
...
...
...
packer.write(SPAN_ID)
```
etc etc ?

Reply:
nvmd, i see discussion here: https://github.com/DataDog/dd-trace-rb/pull/1165#discussion_r483825973